firefox-ios/Storage/Rust/RustRemoteTabs.swift

```diff
@@ -165,6 +172,207 @@ public class RustRemoteTabs {
            self.store?.registerWithSyncManager()
         }
     }
+
+    // MARK: Remote Command APIs
+    public func addRemoteCommand(deviceId: String, url: URL) {
+        DispatchQueue.global().async { [unowned self] in
+            guard let commandQueue = self.commandQueue else {
+                let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+                self.logger.log(err.description,
+                                level: .warning,
+                                category: .tabs)
+                return
+            }
+            commandQueue
+                .addRemoteCommand(deviceId: deviceId, command: RemoteCommand.closeTab(url: url.absoluteString)) { result in
+                    self.handleCommandQueueErrors(result: result)
+                }
+        }
+    }
+
+    public func removeRemoteCommand(deviceId: String, url: URL) {
+        DispatchQueue.global().async { [unowned self] in
+            guard let commandQueue = self.commandQueue else {
+                let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+                self.logger.log(err.description,
+                                level: .warning,
+                                category: .tabs)
+                return
+            }
+            commandQueue
+                .removeRemoteCommand(deviceId: deviceId,
+                                     command: RemoteCommand.closeTab(url: url.absoluteString)) { result in
+                    self.handleCommandQueueErrors(result: result)
+                }
+        }
+    }
+
+    public func getUnsentCommandUrlsByDeviceId(deviceId: String, completion: @escaping ([String]) -> Void) {
+        DispatchQueue.global().async { [unowned self] in
+            self.getUnsentCommandsByDeviceId(deviceId: deviceId) { commands in
+                let urls = commands.map { item in
+                    switch item.command {
+                    case .closeTab(let url):
+                        return url
+                    }
+                }
+                completion(urls)
+            }
+        }
+    }
+
+    public func setPendingCommandsSent(deviceId: String) {
+        DispatchQueue.global().async { [unowned self] in
+            guard let commandQueue = self.commandQueue else {
+                let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+                self.logger.log(err.description,
+                                level: .warning,
+                                category: .tabs)
+                return
+            }
+
+            self.getUnsentCommandsByDeviceId(deviceId: deviceId) { commands in
+                commandQueue.setPendingCommandsSent(deviceId: deviceId, commands: commands) { errors in
```

Comment:
Hmmm, this is assuming that we've sent _all_ pending commands to that device ID, and doing another "queue hop"—`setPendingCommandsSent` hides a `DispatchQueue.global().async` dispatch.

I don't think we can guarantee that all commands were successfully sent, just because we were able to request a flush, though: they might need to be split up into multiple FxA commands, where only some are sent successfully.

This would totally be OK to do in a follow-up; but just noting it here as something we'll want to take care of during the Nightly cycle!

Reply:
https://github.com/mozilla-mobile/firefox-ios/pull/22799/commits/6fef9b6166d054b29c61dc030e93b20f7ca85118